WordPress/src/main/java/org/wordpress/android/reader/savedposts/resolver/ReaderSavedPostsResolver.kt

```diff
@@ -0,0 +1,102 @@
+package org.wordpress.android.reader.savedposts.resolver
+
+import android.database.Cursor
+import org.wordpress.android.R
+import org.wordpress.android.datasets.wrappers.ReaderDatabaseWrapper
+import org.wordpress.android.datasets.wrappers.ReaderPostTableWrapper
+import org.wordpress.android.datasets.wrappers.ReaderTagTableWrapper
+import org.wordpress.android.models.ReaderPostList
+import org.wordpress.android.models.ReaderTag
+import org.wordpress.android.models.ReaderTagType.BOOKMARKED
+import org.wordpress.android.provider.query.QueryResult
+import org.wordpress.android.reader.savedposts.JetpackReaderSavedPostsFlag
+import org.wordpress.android.reader.savedposts.ReaderSavedPostsAnalyticsTracker
+import org.wordpress.android.reader.savedposts.ReaderSavedPostsAnalyticsTracker.ErrorType
+import org.wordpress.android.reader.savedposts.provider.ReaderSavedPostsProvider
+import org.wordpress.android.resolver.ContentResolverWrapper
+import org.wordpress.android.ui.prefs.AppPrefsWrapper
+import org.wordpress.android.util.publicdata.WordPressPublicData
+import org.wordpress.android.viewmodel.ContextProvider
+import javax.inject.Inject
+
+class ReaderSavedPostsResolver @Inject constructor(
+    private val jetpackReaderSavedPostsFlag: JetpackReaderSavedPostsFlag,
+    private val queryResult: QueryResult,
+    private val contextProvider: ContextProvider,
+    private val contentResolverWrapper: ContentResolverWrapper,
+    private val readerSavedPostsAnalyticsTracker: ReaderSavedPostsAnalyticsTracker,
+    private val wordPressPublicData: WordPressPublicData,
+    private val appPrefsWrapper: AppPrefsWrapper,
+    private val readerPostTableWrapper: ReaderPostTableWrapper,
+    private val readerTagTableWrapper: ReaderTagTableWrapper,
+    private val readerDatabaseWrapper: ReaderDatabaseWrapper
+) {
+    fun tryGetReaderSavedPosts(onSuccess: () -> Unit, onFailure: () -> Unit) {
+        val isFeatureFlagEnabled = jetpackReaderSavedPostsFlag.isEnabled()
+        if (!isFeatureFlagEnabled) {
+            onFailure()
+            return
+        }
+        val isFirstTry = appPrefsWrapper.getIsFirstTryReaderSavedPostsJetpack()
+        if (!isFirstTry) {
+            onFailure()
+            return
+        }
+
+        readerSavedPostsAnalyticsTracker.trackStart()
+        appPrefsWrapper.saveIsFirstTryReaderSavedPostsJetpack(false)
+        val savedPostsCursor = getReaderSavedPostsCursor()
+        if (savedPostsCursor != null) {
+            val posts = queryResult.getValue<ReaderPostList>(savedPostsCursor) ?: ReaderPostList()
+
+            if (posts.isNotEmpty()) {
+                updateReaderSavedPosts(onSuccess, onFailure, posts)
+            } else {
+                readerSavedPostsAnalyticsTracker.trackFailed(ErrorType.NoUserSavedPostsError)
+                onFailure
```

Comment:
Should we invoke the function here instead, like `onFailure()`?

Reply:
Of course 🤦‍♂️! Changed [here](206ef3efbb7e49d4167d6c95b9eff50d3f47757f). On a side note, I was working yesterday on the PR that adds unit testing to this code, and a failing test was spotting it for me 😄! Unit testing FTW 🙌!